src/java/com/rapleaf/jack/queries/GenericQueryBuilder.java

```diff
@@ -0,0 +1,176 @@
+package com.rapleaf.jack.queries;
+
+import java.io.IOException;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLRecoverableException;
+import java.util.Set;
+
+import com.rapleaf.jack.BaseDatabaseConnection;
+import com.rapleaf.jack.Column;
+import com.rapleaf.jack.Table;
+import com.rapleaf.jack.queries.where_operators.IWhereOperator;
+
+public class GenericQueryBuilder {
+  private static int MAX_CONNECTION_RETRIES = 1;
+
+  private final GenericQuery genericQuery;
+  private final BaseDatabaseConnection dbConnection;
+
+  public GenericQueryBuilder(BaseDatabaseConnection dbConnection, GenericQuery genericQuery) {
+    this.dbConnection = dbConnection;
+    this.genericQuery = genericQuery;
+  }
+
+  public JoinConditionBuilder leftJoin(Table table) {
+    return new JoinConditionBuilder(this, JoinType.LEFT_JOIN, table);
+  }
+
+  public JoinConditionBuilder rightJoin(Table table) {
+    return new JoinConditionBuilder(this, JoinType.RIGHT_JOIN, table);
+  }
+
+  public JoinConditionBuilder innerJoin(Table table) {
+    return new JoinConditionBuilder(this, JoinType.INNER_JOIN, table);
+  }
+
+  GenericQueryBuilder addJoinCondition(JoinCondition joinCondition) {
+    genericQuery.addJoinCondition(joinCondition);
+    return this;
+  }
+
+  public GenericQueryBuilder where(Column column, IWhereOperator operator) {
```

Comment:
I don't know whether it will be nasty or not yet, but that's fine. There must be a way.

The OOP approach has a big drawback. If there are lots of conditions, there will be a large number of parentheses:
```java
.where((a.or(b)).and(    // [a OR b] AND [c OR d] AND [e OR (f AND g)]
       (c.or(d)).and(
       (e.or(f, g)))))
```

I can combine the rails way into a single `WHERE` clause. Constraints separated by comma within the `where()` method have logic `AND`. Within each constraint, any combinations of `OR` and `AND` relations are allowed. This can significantly reduce the number of parentheses, and it seems easier to implement.

```java
.where(a.or(b),          // [a OR b] AND [c OR d] AND [e OR (f AND g)]
       c.or(d),          // AND logic is always represented by comma
       e.or(f, g))
```